lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+	// structure mimicking folders to keep count of events per directory.
+	rootEventDir *eventDir
+	fsEventChan  chan notify.EventInfo
+	// time interval to search for events to be passed to syncthing-core
+	notifyDelay time.Duration
+	// time after which an active event is passed to syncthing-core
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	resetNotifyTimerChan  chan time.Duration
+	inProgress            map[string]struct{}
+	description           string
+	ignores               *ignore.Matcher
+	ignoresUpdate         chan *ignore.Matcher
+	stop                  chan struct{}
+	ignorePerms           bool
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	FsWatchChan() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+}
+
+// Not meant to be changed, but must be changeable for tests
+var (
+	maxFiles       = 512
+	maxFilesPerDir = 128
+)
+
+func NewFsWatcher(cfg config.FolderConfiguration, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &fsWatcher{
+		folderPath:            filepath.Clean(cfg.Path()),
+		notifyModelChan:       make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		fsEventChan:           make(chan notify.EventInfo, maxFiles),
+		notifyDelay:           time.Duration(cfg.NotifyDelayS) * time.Second,
+		notifyTimeout:         notifyTimeout(cfg.NotifyDelayS),
+		notifyTimerNeedsReset: false,
+		inProgress:            make(map[string]struct{}),
+		description:           cfg.Description(),
+		ignores:               ignores,
+		ignoresUpdate:         make(chan *ignore.Matcher),
+		resetNotifyTimerChan:  make(chan time.Duration),
+		stop:                  make(chan struct{}),
+		ignorePerms:           cfg.IgnorePerms,
+	}
+
+	if err := fsWatcher.setupNotifications(); err != nil {
+		l.Warnf(`Folder "%s": Starting FS notifications failed: %s`,
+			fsWatcher.description, err)
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (watcher *fsWatcher) setupNotifications() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, watcher.folderPath) {
+			return true
+		}
+		relPath, _ := filepath.Rel(watcher.folderPath, absPath)
+		return watcher.ignores.ShouldIgnore(relPath)
+	}
+	// if err := notify.WatchWithFilter(filepath.Join(watcher.folderPath, "..."),
+	if err := notify.WatchWithFilter(filepath.Join(watcher.folderPath, "..."),
+		watcher.fsEventChan, absShouldIgnore, watcher.eventMask()); err != nil {
+		notify.Stop(watcher.fsEventChan)
+		close(watcher.fsEventChan)
+		if isWatchesTooFew(err) {
+			err = watchesLimitTooLowError(watcher.description)
+		}
+		return err
+	}
+	l.Infoln(watcher, "Initiated filesystem watcher")
+	return nil
+}
+
+func (watcher *fsWatcher) Serve() {
+	watcher.notifyTimer = time.NewTimer(watcher.notifyDelay)
+	defer watcher.notifyTimer.Stop()
+
+	inProgressItemSubscription := events.Default.Subscribe(
+		events.ItemStarted | events.ItemFinished)
+
+	for {
+		// Detect channel overflow
```

Comment:
Ups, I considered this bullet-proof due to a pretty naive reasoning: if not true: length<maxFiles -> receive >2 events for overflow: length>maxFiles -> receive one event: length>=maxFiles -> if true: overflow detected. But length can never be bigger than maxFiles...
The following approach can produce false positives, but that's marginal. Any other obvious problems?
```
		case event, _ := <-w.fsEventChan:
                        if len(w.fsEventChan) == maxFiles - 1 {
                                [drain and scan entire folder as before]
                        }
			w.newFsEvent(event.Path(), w.eventType(event.Event()))
```